kerfuffle/archiveentry.cpp

```diff
@@ -0,0 +1,193 @@
+//
+// Created by mvlabat on 5/27/16.
+//
+
+#include "archiveentry.h"
+
+namespace Kerfuffle {
+Archive::Entry::Entry(QObject *parent, QString fullPath, QString rootNode)
+    : QObject(parent)
+    , rootNode(rootNode)
+    , compressedSizeIsSet(true)
+    , m_parent(qobject_cast<Entry*>(parent))
+    , m_size(0)
+    , m_compressedSize(0)
+    , m_isDirectory(false)
+    , m_isPasswordProtected(false)
+{
+    if (!fullPath.isEmpty())
+        setFullPath(fullPath);
+}
+
+Archive::Entry::~Entry()
+{
+    clear();
+}
+
+QVector<Archive::Entry*> Archive::Entry::entries()
+{
+    Q_ASSERT(isDir());
+    return m_entries;
+}
+
+const QVector<Archive::Entry*> Archive::Entry::entries() const {
+    Q_ASSERT(isDir());
+    return m_entries;
+}
+
+void Archive::Entry::setEntryAt(int index, Entry *value)
+{
+    Q_ASSERT(isDir());
+    Q_ASSERT(index < m_entries.count());
+    m_entries[index] = value;
+}
+
+void Archive::Entry::appendEntry(Entry *entry)
+{
+    Q_ASSERT(isDir());
+    m_entries.append(entry);
+}
+
+void Archive::Entry::removeEntryAt(int index)
+{
+    Q_ASSERT(isDir());
+    Q_ASSERT(index < m_entries.count());
+    delete m_entries.takeAt(index);
+}
+
+Archive::Entry *Archive::Entry::getParent() const
+{
+    return m_parent;
+}
+
+void Archive::Entry::setParent(Archive::Entry *parent)
+{
+    m_parent = parent;
+}
+
+void Archive::Entry::setFullPath(const QString &fullPath)
+{
+    m_fullPath = fullPath;
+    processName();
+    processIcon();
+}
+
+void Archive::Entry::setIsDirectory(const bool isDirectory)
+{
+    m_isDirectory = isDirectory;
+    processName();
+    processIcon();
+}
+
+int Archive::Entry::row() const
+{
+    if (getParent()) {
+        return getParent()->entries().indexOf(const_cast<Archive::Entry*>(this));
+    }
+    return 0;
+}
+
+bool Archive::Entry::isDir() const
+{
+    return m_isDirectory;
+}
+
+void Archive::Entry::processName()
+{
+    const QStringList pieces = m_fullPath.split(QLatin1Char( '/' ), QString::SkipEmptyParts);
+    m_name = pieces.isEmpty() ? QString() : pieces.last();
+}
+
+void Archive::Entry::processIcon()
+{
+    QMimeDatabase db;
+    if (isDir()) {
+        m_icon = QIcon::fromTheme(db.mimeTypeForName(QStringLiteral("inode/directory")).iconName()).pixmap(IconSize(KIconLoader::Small),
+                                                                                                           IconSize(KIconLoader::Small));
+    } else {
+        m_icon = QIcon::fromTheme(db.mimeTypeForFile(m_fullPath).iconName()).pixmap(IconSize(KIconLoader::Small),
```

Comment:
This could be the reason why `JobsTest` crashes. The icon returned by `QIcon::fromTheme()` might be "null" and you should check this with the `QIcon::isNull()` method.